pkg/config/resource.go

```diff
@@ -20,40 +20,59 @@ import (
+// SetIdentifierArgumentsFn sets the name of the resource in Terraform attributes map,
+// i.e. Main HCL file.
+type SetIdentifierArgumentsFn func(base map[string]interface{}, name string)
 
 // NopSetIdentifierArgument does nothing. It's useful for cases where the external
 // name is calculated by provider and doesn't have any effect on spec fields.
-func NopSetIdentifierArgument(_ map[string]interface{}, _ string) {}
+var NopSetIdentifierArgument SetIdentifierArgumentsFn = func(_ map[string]interface{}, _ string) {}
+
+// GetIDFn returns the ID to be used in TF State file, i.e. "id" field in
+// terraform.tfstate.
+type GetIDFn func(name string, parameters map[string]interface{}, providerConfig map[string]interface{}) string
+
+// NameAsID returns the name to be used as ID in TF State file.
+var NameAsID GetIDFn = func(name string, _ map[string]interface{}, _ map[string]interface{}) string {
+	return name
+}
+
+// GetNameFn returns the external name extracted from the TF State.
+type GetNameFn func(tfstate map[string]interface{}) string
+
+// IDAsName returns the TF State ID as external name.
+var IDAsName GetNameFn = func(tfstate map[string]interface{}) string {
+	if id, ok := tfstate["id"].(string); ok {
+		return id
+	}
+	return ""
+}
 
 // AdditionalConnectionDetailsFn functions adds custom keys to connection details
 // secret using input terraform attributes
 type AdditionalConnectionDetailsFn func(attr map[string]interface{}) (map[string][]byte, error)
 
 // NopAdditionalConnectionDetails does nothing, when no additional connection
 // details configuration function provided.
-func NopAdditionalConnectionDetails(_ map[string]interface{}) (map[string][]byte, error) {
+var NopAdditionalConnectionDetails AdditionalConnectionDetailsFn = func(_ map[string]interface{}) (map[string][]byte, error) {
 	return nil, nil
 }
 
-// ResourceOption allows setting optional fields of a Resource object.
-type ResourceOption func(*Resource)
-
-// WithTerraformIDFieldName allows you to set IDFieldName.
-func WithTerraformIDFieldName(n string) ResourceOption {
-	return func(c *Resource) {
-		c.IDFieldName = n
-	}
-}
-
 // ExternalName contains all information that is necessary for naming operations,
 // such as removal of those fields from spec schema and calling Configure function
 // to fill attributes with information given in external name.
 type ExternalName struct {
 	// SetIdentifierArgumentFn sets the name of the resource in Terraform argument
-	// map.
-	SetIdentifierArgumentFn SetIdentifierArgumentFn
+	// map, otherwise we cannot know which field in HCL we should treat as identifier.
+	SetIdentifierArgumentFn SetIdentifierArgumentsFn
+
+	// GetNameFn returns the external name extracted from TF State. In most cases,
+	// "id" key of TF State should be returned.
+	GetNameFn GetNameFn
+
+	// GetIDFn returns the string that will be used as "id" key in TF state. In
+	// most cases, it should return given external name as is.
+	GetIDFn GetIDFn
```

Comment:
nit: The immediate context here is (the enclosing type) the `ExternalName` struct, which is a Crossplane concept. I think it will be more explicit to rename this field as `GetTerraformIDFn`.

Reply:
I'd actually prefer to keep it as ID since that's what it's called in the map that this function processes, similarly how `ExternalName` corresponds to `crossplane.io/external-name` annotation.